File: apple/Packages/FtpAnimeCore/Sources/FtpAnimeCore/ApiClient.swift
import Foundation

/// Typed async wrapper around the Spring Boot REST API. Every call uses
/// HTTP Basic auth (same credentials as the existing sync endpoint).
///
/// Backend URL and credentials are stored outside this type — the caller
/// builds it with a `Configuration` and then the `@Published` state in
/// `LibraryStore` keeps the UI in sync.
public actor ApiClient {

    public struct Configuration: Equatable, Sendable {
        public var baseURL: URL
        public var username: String
        public var password: String

        public init(baseURL: URL, username: String, password: String) {
            self.baseURL = baseURL
            self.username = username
            self.password = password
        }
    }

    public enum ApiError: LocalizedError {
        case badStatus(Int, String?)
        case decode(Error)
        case transport(Error)

        public var errorDescription: String? {
            switch self {
            case .badStatus(let code, let body): return "HTTP \(code)\(body.map { ": \($0)" } ?? "")"
            case .decode(let err): return "decode failed: \(err.localizedDescription)"
            case .transport(let err): return err.localizedDescription
            }
        }
    }

    private let configuration: Configuration
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    public init(configuration: Configuration, session: URLSession = .shared) {
        self.configuration = configuration
        self.session = session
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    public var baseURL: URL { configuration.baseURL }

    // MARK: - servers

    public func listServers() async throws -> [FtpServerConfig] {
        try await get("/api/servers")
    }

    public func upsertServer(_ cfg: FtpServerConfig) async throws -> FtpServerConfig {
        try await post("/api/servers", body: cfg)
    }

    public func removeServer(id: String) async throws {
        _ = try await request(method: "DELETE", path: "/api/servers/\(id)")
    }

    public func testServer(_ cfg: FtpServerConfig) async throws -> ConnectionTestResult {
        try await post("/api/servers/test", body: cfg)
    }

    public func browse(serverId: String, path: String) async throws -> [RemoteEntry] {
        try await get("/api/servers/\(serverId)/browse", query: ["path": path])
    }

    // MARK: - library

    public func listLibraryRoots() async throws -> [LibraryRoot] {
        try await get("/api/library-roots")
    }

    public func addLibraryRoot(_ root: LibraryRoot) async throws -> LibraryRoot {
        try await post("/api/library-roots", body: root)
    }

    public func removeLibraryRoot(id: String) async throws {
        _ = try await request(method: "DELETE", path: "/api/library-roots/\(id)")
    }

    public func cachedLibrary() async throws -> [AnimeEntry] {
        try await get("/api/library")
    }

    public func loadAnime(serverId: String, path: String, libraryRootId: String) async throws -> AnimeEntry {
        try await get("/api/library/anime", query: [
            "serverId": serverId,
            "path": path,
            "libraryRootId": libraryRootId
        ])
    }

    // MARK: - metadata

    public func searchMetadata(query: String) async throws -> [AnimeMetadata] {
        try await get("/api/metadata/search", query: ["q": query])
    }

    // MARK: - streaming

    public struct RegisterRequest: Codable, Sendable {
        public var serverId: String
        public var path: String
        public var size: Int64
        public init(serverId: String, path: String, size: Int64) {
            self.serverId = serverId; self.path = path; self.size = size
        }
    }

    public func registerStream(_ req: RegisterRequest) async throws -> StreamHandle {
        try await post("/api/stream/register", body: req)
    }

    public func unregisterStream(token: String) async throws {
        _ = try await request(method: "DELETE", path: "/api/stream/\(token)")
    }

    public func probe(token: String) async throws -> ProbeResult {
        try await get("/api/stream/\(token)/probe")
    }

    // MARK: - progress / favorites

    public func getProgress(serverId: String, path: String) async throws -> VideoProgress? {
        let data = try await request(method: "GET", path: "/api/progress", query: ["serverId": serverId, "path": path])
        if data.isEmpty { return nil }
        return try decode(data)
    }

    public func setProgress(_ progress: VideoProgress) async throws {
        _ = try await request(method: "POST", path: "/api/progress", body: progress)
    }

    public func listProgress() async throws -> [VideoProgress] {
        try await get("/api/progress/all")
    }

    public func listFavorites() async throws -> [FavoriteFolder] {
        try await get("/api/favorites")
    }

    public func addFavorite(_ favorite: FavoriteFolder) async throws -> [FavoriteFolder] {
        try await post("/api/favorites", body: favorite)
    }

    public func removeFavorite(serverId: String, path: String) async throws -> [FavoriteFolder] {
        let data = try await request(method: "DELETE", path: "/api/favorites", query: ["serverId": serverId, "path": path])
        return try decode(data)
    }

    // MARK: - URL builders

    /// Produce a fully-qualified playback URL for AVPlayer. Includes HTTP
    /// Basic credentials via `URLComponents.user/.password` so AVPlayer
    /// hits the Spring Boot endpoint without needing a custom loader.
    /// `nonisolated` because this only reads the immutable `configuration`.
    public nonisolated func playbackURL(for handle: StreamHandle, preferDirect: Bool) -> URL? {
        let rel = (preferDirect ? handle.directUrl : (handle.hlsUrl ?? handle.transcodeUrl))
        guard var components = URLComponents(url: configuration.baseURL, resolvingAgainstBaseURL: false) else { return nil }
        let joined = rel.hasPrefix("/") ? rel : "/\(rel)"
        components.path = joined
        components.user = configuration.username
        components.password = configuration.password
        return components.url
    }

    public nonisolated func posterURL(path: String) -> URL? {
        guard var components = URLComponents(url: configuration.baseURL, resolvingAgainstBaseURL: false) else { return nil }
        let name = (path as NSString).lastPathComponent
        components.path = "/api/sync/cache/posters/\(name)"
        components.user = configuration.username
        components.password = configuration.password
        return components.url
    }

    // MARK: - internals

    private func get<T: Decodable>(_ path: String, query: [String: String] = [:]) async throws -> T {
        let data = try await request(method: "GET", path: path, query: query)
        return try decode(data)
    }

    private func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        let data = try await request(method: "POST", path: path, body: body)
        return try decode(data)
    }

    private func decode<T: Decodable>(_ data: Data) throws -> T {
        do { return try decoder.decode(T.self, from: data) }
        catch { throw ApiError.decode(error) }
    }

    private func request(method: String, path: String, query: [String: String] = [:], body: Encodable? = nil) async throws -> Data {
        guard var components = URLComponents(url: configuration.baseURL, resolvingAgainstBaseURL: false) else {
            throw ApiError.transport(URLError(.badURL))
        }
        components.path = path
        if !query.isEmpty {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        guard let url = components.url else { throw ApiError.transport(URLError(.badURL)) }
        var request = URLRequest(url: url)
        request.httpMethod = method
        let credential = "\(configuration.username):\(configuration.password)"
        if let auth = credential.data(using: .utf8)?.base64EncodedString() {
            request.setValue("Basic \(auth)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = try encoder.encode(AnyEncodable(body))
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw ApiError.transport(URLError(.badServerResponse))
            }
            if http.statusCode == 404 { return Data() }
            if !(200..<300).contains(http.statusCode) {
                let body = String(data: data, encoding: .utf8)
                throw ApiError.badStatus(http.statusCode, body)
            }
            return data
        } catch let apiError as ApiError {
            throw apiError
        } catch {
            throw ApiError.transport(error)
        }
    }
}

/// Erases a concrete `Encodable` so the generic `request` signature stays
/// simple; Swift doesn't allow `any Encodable` in all the places we'd want.
private struct AnyEncodable: Encodable {
    let wrapped: Encodable

    init(_ wrapped: Encodable) { self.wrapped = wrapped }

    func encode(to encoder: Encoder) throws {
        try wrapped.encode(to: encoder)
    }
}

File: android/core/src/main/kotlin/com/ftpanimeplayer/core/api/ApiClient.kt
package com.ftpanimeplayer.core.api

import kotlinx.serialization.json.Json
import okhttp3.Credentials
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.HTTP
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query
import java.util.concurrent.TimeUnit

/**
 * Retrofit-based client for the Spring Boot backend. Same endpoints as the
 * iOS/macOS clients. Auth uses HTTP Basic, attached by an OkHttp interceptor.
 *
 * Construct via [ApiClient.create]; the singleton nature is the caller's
 * responsibility (typically held in a ViewModel or Application).
 */
class ApiClient private constructor(
    val config: Configuration,
    val http: OkHttpClient,
    private val service: FtpAnimeService
) : FtpAnimeService by service {

    data class Configuration(val baseUrl: String, val username: String, val password: String)

    /** Absolute URL for an HLS master.m3u8 / direct stream / poster. */
    fun playbackUrl(handle: StreamHandle, preferDirect: Boolean): String {
        val rel = if (preferDirect) handle.directUrl else (handle.hlsUrl ?: handle.transcodeUrl)
        return joinUrl(config.baseUrl, rel)
    }

    fun posterUrl(posterPath: String): String {
        val name = posterPath.substringAfterLast('/')
        return joinUrl(config.baseUrl, "/api/sync/cache/posters/$name")
    }

    companion object {
        fun create(config: Configuration): ApiClient {
            val authInterceptor = Interceptor { chain ->
                val creds = Credentials.basic(config.username, config.password)
                val req = chain.request().newBuilder()
                    .header("Authorization", creds)
                    .header("Accept", "application/json")
                    .build()
                chain.proceed(req)
            }
            val http = OkHttpClient.Builder()
                .addInterceptor(authInterceptor)
                .connectTimeout(20, TimeUnit.SECONDS)
                // Long read timeout because SSE library scans can stay open
                // for minutes while the backend enriches metadata.
                .readTimeout(10, TimeUnit.MINUTES)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build()
            val json = Json {
                ignoreUnknownKeys = true
                explicitNulls = false
            }
            val retrofit = Retrofit.Builder()
                .baseUrl(ensureTrailingSlash(config.baseUrl).toHttpUrl())
                .client(http)
                .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
                .build()
            val service = retrofit.create(FtpAnimeService::class.java)
            return ApiClient(config, http, service)
        }

        private fun joinUrl(baseUrl: String, path: String): String {
            val trimmedBase = baseUrl.trimEnd('/')
            val leading = if (path.startsWith("/")) path else "/$path"
            return "$trimmedBase$leading"
        }

        private fun ensureTrailingSlash(url: String) = if (url.endsWith("/")) url else "$url/"
    }
}

interface FtpAnimeService {
    @GET("api/servers") suspend fun listServers(): List<FtpServerConfig>
    @POST("api/servers") suspend fun upsertServer(@Body cfg: FtpServerConfig): FtpServerConfig
    @DELETE("api/servers/{id}") suspend fun removeServer(@Path("id") id: String)
    @POST("api/servers/test") suspend fun testServer(@Body cfg: FtpServerConfig): ConnectionTestResult
    @GET("api/servers/{id}/browse") suspend fun browse(@Path("id") id: String, @Query("path") path: String): List<RemoteEntry>

    @GET("api/library-roots") suspend fun listLibraryRoots(): List<LibraryRoot>
    @POST("api/library-roots") suspend fun addLibraryRoot(@Body root: LibraryRoot): LibraryRoot
    @DELETE("api/library-roots/{id}") suspend fun removeLibraryRoot(@Path("id") id: String)

    @GET("api/library") suspend fun cachedLibrary(): List<AnimeEntry>

    @GET("api/library/anime") suspend fun loadAnime(
        @Query("serverId") serverId: String,
        @Query("path") path: String,
        @Query("libraryRootId") libraryRootId: String
    ): AnimeEntry

    @GET("api/metadata/search") suspend fun searchMetadata(@Query("q") query: String): List<AnimeMetadata>

    @POST("api/stream/register") suspend fun registerStream(@Body req: RegisterStreamRequest): StreamHandle
    @DELETE("api/stream/{token}") suspend fun unregisterStream(@Path("token") token: String)
    @GET("api/stream/{token}/probe") suspend fun probe(@Path("token") token: String): ProbeResult

    @GET("api/progress") suspend fun getProgress(@Query("serverId") serverId: String, @Query("path") path: String): VideoProgress?
    @POST("api/progress") suspend fun setProgress(@Body progress: VideoProgress)
    @GET("api/progress/all") suspend fun listProgress(): List<VideoProgress>

    @GET("api/favorites") suspend fun listFavorites(): List<FavoriteFolder>
    @POST("api/favorites") suspend fun addFavorite(@Body fav: FavoriteFolder): List<FavoriteFolder>
    @HTTP(method = "DELETE", path = "api/favorites")
    suspend fun removeFavorite(@Query("serverId") serverId: String, @Query("path") path: String): List<FavoriteFolder>
}
